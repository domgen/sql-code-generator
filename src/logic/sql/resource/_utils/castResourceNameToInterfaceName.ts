import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
