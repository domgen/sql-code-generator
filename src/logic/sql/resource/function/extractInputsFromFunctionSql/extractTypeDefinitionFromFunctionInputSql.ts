import { DataType } from '../../../../../model';
import { TypeDefinitionOfResourceInput } from '../../../../../model/valueObjects/TypeDefinitionOfResourceInput';
import { extractDataTypeFromColumnOrArgumentDefinitionSql } from '../../common/extractDataTypeFromColumnOrArgumentDefinitionSql';

export const extractTypeDefinitionFromFunctionInputSql = ({ sql }: { sql: string }) => {
  // 1. extract the name; its typically the first string
  const name = sql.split(' ')[0].replace(/[^a-zA-Z_]+/gi, '');

  // 2. extract the root type;
  const primaryType = extractDataTypeFromColumnOrArgumentDefinitionSql({ sql });

  // 3. define the full type definition; note: array => union
  const type = [primaryType].filter((type) => !!type) as DataType[];

  // 4. return the definition
  return new TypeDefinitionOfResourceInput({
    name,
    type,
  });
};
