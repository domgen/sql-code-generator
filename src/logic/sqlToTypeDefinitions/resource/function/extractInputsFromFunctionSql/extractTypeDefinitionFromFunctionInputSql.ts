import { DataType } from '../../../../../domain';
import { TypeDefinitionOfResourceInput } from '../../../../../domain/objects/TypeDefinitionOfResourceInput';
import { extractDataTypeFromColumnOrArgumentDefinitionSql } from '../../common/extractDataTypeFromColumnOrArgumentDefinitionSql';

export const extractTypeDefinitionFromFunctionInputSql = ({
  sql,
}: {
  sql: string;
}) => {
  // 1. extract the name; its typically the first string
  const name = sql.split(' ')[0]!.replace(/[^a-zA-Z_]+/gi, '');

  // 2. extract the root type;
  const primaryType = extractDataTypeFromColumnOrArgumentDefinitionSql({ sql });

  // 3. define the full type definition; note: array => union
  const type = [
    primaryType,
    DataType.NULL, // note: inputs to functions are always nullable ?
  ].filter((type) => !!type) as DataType[];

  // 4. return the definition
  return new TypeDefinitionOfResourceInput({
    name,
    type,
  });
};
