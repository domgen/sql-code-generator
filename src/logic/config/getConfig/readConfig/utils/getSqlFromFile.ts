import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
