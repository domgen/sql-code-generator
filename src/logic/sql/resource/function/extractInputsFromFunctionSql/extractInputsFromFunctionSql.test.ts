import { extractInputsFromFunctionSql } from './extractInputsFromFunctionSql';
import { getSqlFromFile } from '../../../../config/_utils/getSqlFromFile';

describe('extractInputsFromFunctionSql', () => {
  it('should extract the inputs accurately in this example', async () => {
    const defs = extractInputsFromFunctionSql({
      sql: await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/upsert_image.sql` }),
    });
    expect(defs).toMatchSnapshot();
  });
  it('should extract the inputs accurately in this other example', async () => {
    const defs = extractInputsFromFunctionSql({
      sql: await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/upsert_suggestion.sql` }),
    });
    expect(defs).toMatchSnapshot();
  });
  it('should extract the inputs accurately in this other example again', async () => {
    const defs = extractInputsFromFunctionSql({
      sql: await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/hash_string.sql` }),
    });
    expect(defs).toMatchSnapshot();
  });
});
