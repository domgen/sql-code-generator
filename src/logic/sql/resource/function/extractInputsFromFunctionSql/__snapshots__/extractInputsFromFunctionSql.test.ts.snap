// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractInputsFromFunctionSql should extract the inputs accurately in this example 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_url",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_caption",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_credit",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_alt_text",
    "type": Array [
      "string",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql should extract the inputs accurately in this other example 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_suggestion_source",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_external_id",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_suggested_idea_id",
    "type": Array [
      "number",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_status",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_result",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_resultant_curated_idea_id",
    "type": Array [
      "number",
      "null",
    ],
  },
]
`;

exports[`extractInputsFromFunctionSql should extract the inputs accurately in this other example again 1`] = `
Array [
  TypeDefinitionOfResourceInput {
    "name": "in_array",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_delimiter",
    "type": Array [
      "string",
      "null",
    ],
  },
  TypeDefinitionOfResourceInput {
    "name": "in_index",
    "type": Array [
      "number",
      "null",
    ],
  },
]
`;
